XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+
+#### `init(id, type, parentId)`
+
+* `id` {Number}
+* `type` {String}
+* `parentId` {Number}
+
+Called when a class is constructed that has the possibility to trigger an
+asynchronous event. This does mean the instance _will_ trigger a
+`pre()`/`post()` event. Only that the possibility exists.
+
+The `this` of each call is that of the object being constructed. While core
+attempts to make accessing properties of these handles safe at construction, it
+is not guaranteed (especially if third-party modules are involved) that
+everything can be safely accessed. So it is not recommended to blindly
+iterate over the object's properties. Despite this, it is important to give
+access of the handle to the user so they can utilize tracking methods of their
+choosing. For example:
+
+```js
+const handleMap = new Map();
+
+function HandleInst(inst) {
+  // Emulate performance.now()
+  const t = process.hrtime();
+  this.time = t[0] * 1e3 + t[1] / 1e6;
+  this.inst = inst;
+}
+
+function init(id) {
+  handleMap.set(id, new HandleInst(this));
+}
+```
+
+Every instance is assigned a unique id. Taking advantage of the fraction space
+in the 64-bit IEEE 754-2008 that ECMAScript defines as the **Number** type, the
+number of id's that can be assigned are `2^53 - 1` (also defined as
+`Number.MAX_SAFE_INTEGER`). At this size node can assign a new id every 100
+nanoseconds and not run out for over 28 years. Because of this circumstance it
+is not deemed necessary to find an alternative approach.
+
+The id of the parent in the async stack trace is also passed. This can be used
+to create call graphs, trace resource usage/leakage, etc.
+
+#### `pre(id)`
+
+* `id` {Number}
+
+Called just before the return callback is called after completing an
+asynchronous request. Or called on handles with events such as receiving a new
+connection. For requests, such as `fs.open()`, this should be called exactly
+once. For handles, such as a TCP server, this may be called 0-N times. Useful
+for recording metrics or setting up state.
+
+#### `post(id, didThrow)`
+
+* `id` {Number}
+* `didThrow` {Boolean}
+
+Called immediately after the return callback is completed. If the callback
+threw but was caught by a domain or `uncaughtException` `didThrow` will be set
+to `true`. If the callback threw but was not caught then the process will exit
+immediately without calling `post()`.
+
+#### `destroy(id)`
+
+* `id` {Number}
+
+Called either when the class destructor is run, or if the resource is marked as
+free. The destructor will usually run when explicitly called (the case for
+handles) or when a request has completed. In at least one case this will be
+triggered by GC. In the case of shared or cached resources, such as
+`HTTPParser`, `destroy()` will be manually called when the TCP connection no
+longer needs it.
+
+Because the callback is called during deconstruction of the class instance it
+is not safe to access the JS handle during `destroy()` execution. Reason this
```

Comment:
Yes. To reiterate a comment I just posted as to why this is the case:

> Now, destroy() doesn't have a this because it's essentially a static method. It's unsafe to access a C++ class instance while deconstructing. So only the id of the handle is passed to JS so any storage resources can be released.

So you could essentially see `undefined` as the handle instance at that point in time. Since it has been (or soon will be) free'd.